File: ios/Modules/Sources/VortexCrypto/api/Kdf.swift
import Foundation

/// HKDF-SHA256 (RFC 5869).
///
/// Narrow by design — password-based derivation lives in [PasswordHasher]
/// because it needs memory/iteration knobs HKDF doesn't carry.
public protocol Kdf: Sendable {
    func derive(ikm: Data, salt: Data, info: Data, length: Int) throws -> Data
}

public extension Kdf {
    func derive(ikm: Data, salt: Data = Data(), info: Data = Data(), length: Int = 32) throws -> Data {
        try derive(ikm: ikm, salt: salt, info: info, length: length)
    }
}

public enum KdfError: Error {
    case outputTooLong(requested: Int, max: Int)
}

File: android/app/src/main/java/sol/vortexx/android/crypto/api/Kdf.kt
package sol.vortexx.android.crypto.api

/**
 * Key derivation function — HKDF-SHA256 per RFC 5869.
 *
 * Kept narrow on purpose: `derive(ikm, salt, info, len)`. Password-based
 * derivation is a different shape (adds memory/iteration knobs) and lives
 * in [PasswordHasher].
 */
interface Kdf {
    fun derive(
        ikm: ByteArray,
        salt: ByteArray = ByteArray(0),
        info: ByteArray = ByteArray(0),
        length: Int = 32,
    ): ByteArray
}
